src/frontend/src/optimizer/plan_node/stream_vector_index_write.rs

```diff
@@ -0,0 +1,266 @@
+    }
+
+    pub fn create(
+        StreamOptimizedLogicalPlanRoot {
+            plan: input,
+            required_order: user_order_by,
+            out_fields: user_cols,
+            out_names,
+            ..
+        }: StreamOptimizedLogicalPlanRoot,
+        name: String,
+        database_id: DatabaseId,
+        schema_id: SchemaId,
+        definition: String,
+        cardinality: Cardinality,
+        retention_seconds: Option<NonZeroU32>,
+        vector_index_info: PbVectorIndexInfo,
+    ) -> crate::error::Result<Self> {
+        let input = RequiredDist::PhysicalDist(Distribution::Single)
+            .streaming_enforce_if_not_satisfies(input)?;
+        // the hidden column name might refer some expr id
+        let input = reorganize_elements_id(input);
+        let columns = derive_columns(input.schema(), out_names, &user_cols)?;
+
+        let create_type = if input.ctx().session_ctx().config().background_ddl()
+            && plan_can_use_background_ddl(&input)
+        {
+            CreateType::Background
+        } else {
+            CreateType::Foreground
+        };
+
+        let table = Self::derive_table_catalog(
+            input.clone(),
+            name,
+            database_id,
+            schema_id,
+            user_order_by,
+            columns,
+            definition,
+            cardinality,
+            retention_seconds,
+            create_type,
+            vector_index_info,
+        )?;
+
+        Self::new(input, table)
+    }
+
+    #[expect(clippy::too_many_arguments)]
+    fn derive_table_catalog(
+        rewritten_input: PlanRef,
+        name: String,
+        database_id: DatabaseId,
+        schema_id: SchemaId,
+        user_order_by: Order,
+        columns: Vec<ColumnCatalog>,
+        definition: String,
+        cardinality: Cardinality,
+        retention_seconds: Option<NonZeroU32>,
+        create_type: CreateType,
+        vector_index_info: PbVectorIndexInfo,
+    ) -> crate::error::Result<TableCatalog> {
+        let input = rewritten_input;
+
+        let value_indices = (0..columns.len()).collect_vec();
+        assert_eq!(input.distribution(), &Distribution::Single);
+        let distribution_key = vec![];
+        let append_only = input.append_only();
+        // TODO(rc): In `TableCatalog` we still use `FixedBitSet` for watermark columns, ignoring the watermark group information.
+        // We will record the watermark group information in `TableCatalog` in the future. For now, let's flatten the watermark columns.
+        let watermark_columns = input.watermark_columns().indices().collect();
+
+        let (table_pk, stream_key) = derive_pk(input, user_order_by, &columns);
+        // assert: `stream_key` is a subset of `table_pk`
+
+        let read_prefix_len_hint = table_pk.len();
+        // We don't need to fill in table id for table in frontend. The id will be generated on
+        // meta catalog service.
+        Ok(TableCatalog {
```

Comment:
Shall we use TableCatalogBuilder instead?

Reply:
I had a try, but it seems that the functionality of the builder is limiter, so I build the table catalog manually. Most code is copied from the `StreamMaterialize::derive_table_catalog`.